### ISSUE TITLE
Class inheritance of Data reduces code readability

### ISSUE THREAD

**Issue by sguthrie** [sguthrie/osl-dynamics#1]
`RW`, `Processing`, and `TensorFlowDataset` are only used in the inheritance of `Data`. They assume the existence of variables that are defined in the inheriting class. This seems to be a sub-optimal data structure, and ought to be improved

**Comment by cgohil8** [comment]
Yeah, originally it was one (very) big class. We switched to a Mixin class to make the code more modular. I have been thinking about going back to one big class, which would simplify the documentation. I might implement this in the future if it makes sense to do so.

**Comment by sguthrie** [comment]
@cgohil8 I think separating into classes could definitely increase readability. But the current implementation seems to overload `Data` with multiple concepts. 

It seems like there are several types of `Data` (ex: a set of raw data, tde-processed data, amplitude-envelope processed data) that have some constant attributes and methods.
  * Each Data object has a set of subjects or subject data, `n_subjects`, `n_samples`, `n_channels`
  * Each Data object has a place it's being stored
  * Each Data object should have a way to save itself to its storage location
  * Each Data object should have a way to clean up the storage directory 
  * Each Data object can be returned as a time series
  * Each Data object can return a TensorFlow dataset

What do you think of creating an AbstractBaseClass for this concept? Inheritance might look like the following:
  - `RawData` (implementing a `load_data`) function
  - `ProcessedData` (another AbstractBaseClass including the attributes `n_embeddings`, `source_raw_data`, implementing a `save` method, and defining an abstract method `process`)
    - `TdeProcessedData`
    - `AmplitudeEnvelopeProcessedData`

To hold these all together (replicating what's done currently) a class `DataCollection` could replace `Data`. Properties would include:
    * `_identifier`
    * `inputs` (the inputs given by the user to initialize the collection)
    * `raw_data_obj` (an instance of `RawData`)
    * `processed_data_obj` (an instance of `ProcessedData`)
    * `processed` (or `prepared`)
Functions would include:
    * `process_raw_data` - sets the `processed_data_obj` based on the keyword arguments

I appreciate this separation of data into raw and processed since it requires a user to be explicit about what they're doing. Or at least it provides a way in the code to clearly show the rules about which data are being used to produce a Tensorflow dataset. This mechanism also means the number of IDE flags goes down, which means bugs can be found more easily.

I've created a branch with a rudimentary refactor as I was going through the code learning it, in case it might be helpful to you. Before creating a PR to the original repo, I'd definitely write some tests on main and make sure they pass here too :)
